### PR TITLE
Associations

### PR DESCRIPTION
**Pull request by danielamoreno699**
**_In this milestone project, I :_**

-Created a class Classroom with the following requirements: @label instance variable, should be initialized in the constructor.
Setter and getter for @label (remember about attr_accessor).

-Created the has-many/belongs-to relationship between Classroom and Student. The following should be implemented:
Create the has-many side (a classroom has many students). Create the belongs-to side (a student belongs to a classroom).
Make sure that when adding a student to a classroom it also sets the classroom for the student. Make sure that when setting the classroom for a student it also adds it to the classrooms' students.

-Created a class Book with the following:@title and @author instance variables, should be initialized in the constructor.
Setters and getters for instance variables (remember about attr_accessor).

-Created a class Rental with the following:@date instance variable, should be initialized in the constructor.
Setter and getter for @date (remember about attr_accessor).

-Created the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental. The following should be implemented: Create the has-many side of Book and Rental (a book has many rentals).Create the belongs-to side of Rental and Book (a rental belongs to a book).Create the has-many side of Person and Rental (a person has many rentals).Create the belongs-to side of Rental and Person (a rental belongs to a person).

-Modified the constructor of Rental so Book and Person are set in it.

### CODE REVIEW

File: book.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class Book
+  attr_accessor :title, :author
+  attr_reader :rentals
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+end
```
Comment (cynthiainga) [inline book.rb]: - [ ] You should have a method for adding rentals, the book itself should be sent as a parameter to create the new rental achieving the `has-many` association. Please kindly add that method 🙃 
- [ ] Also `Getters` and `Setters` for `@rentals` should be present as mentioned in the project which means `attr_accessor` and not `attr_reader`, `attr_reader` is just for `setters` 😁 😁

File: rental.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+class Rental
+  attr_accessor :date
+  attr_reader :book, :person
```
Comment (cynthiainga) [inline rental.rb]: - [ ] `Getters and Setters` for `@book` and `@person` instance variables should be present which means `attr_accessor` and not `attr_reader` for book and person, kindly fix that 🙃 😁

File: person.rb [hunk 3]
```diff
@@ -3,14 +3,15 @@
 class Person < Nameable
   # setters and getters
   attr_accessor :name, :age
-  attr_reader :id
+  attr_reader :id, :rentals
```
Comment (cynthiainga) [inline person.rb]: - [ ] `Getters and Setters` for `@rentals` should be present which means `attr_accessor` and not `attr_reader`, please kindly fix that 🙏🏾 
- [ ] Also you should have a method for adding rentals should be present, and the person itself should be sent as a parameter to create the new rental achieving the `has-many` association, kindly add the method 🙃 😁

**Review by cynthiainga — CHANGES_REQUESTED** [review]
Hi @danielamoreno699, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights

- [ ] A rental belongs to a book and also belongs to a person ✔️ 
- [ ] The `has-many` / `belongs-to` relationships between `Classroom` and `Student` are created ✔️ 
- [ ] Correct Gitflow ✔️ 

## Required Changes ♻️

- [ ] Please kindly check the comments under the review 👇🏾 👇🏾 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please remember to tag me in your question so I can get the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by danielamoreno699** [comment]
@cynthiainga hi cynthia, Im a bit confused. you are asking me to add a method into the books. to add the books into rentals 
but the book is already added by using this method '
    @book = book
    book.rentals << self
' to requirement was to Created the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental. and the class should be the one that adds persons and books. I already tested in the irb and works 

'Also Getters and Setters for @rentals should be present as mentioned in the project which means attr_accessor and not attr_reader, attr_reader is just for setters' this review is a bit confusing. becasue the attr_accesor is a setter and the attr_reader is a getter. you are saying that attr_accesor is a getter and setter at the same time but this is not true

File: book.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_book_to_rentals(book)
+    @rentals.push(book)
+    book.rentals = self
+  end
```
Comment (danmainah) [inline book.rb]: - [ ] @danielamoreno699 great work implementing the changes requested, however, you did it wrongly, as the previous reviewer requested you add the `add_rental` method that takes in the `person, date, and book itself`, this enables us to achieve the 'has-many' association` between the person and book classes. 

`hint`
```
def add_rental(person, date) 
    Rental.new(date, self, person)
  end
```

**Comment by danmainah** [comment]
> @cynthiainga hi cynthia, Im a bit confused. you are asking me to add a method into the books. to add the books into rentals but the book is already added by using this method ' @book = book book.rentals << self ' to requirement was to Created the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental. and the class should be the one that adds persons and books. I already tested in the irb and works
> 
> 'Also Getters and Setters for @rentals should be present as mentioned in the project which means attr_accessor and not attr_reader, attr_reader is just for setters' this review is a bit confusing. becasue the attr_accesor is a setter and the attr_reader is a getter. you are saying that attr_accesor is a getter and setter at the same time but this is not true

@danielamoreno699 Thank you for your message, as the earlier reviewer pointed out the method of adding rental, is supposed to achieve the has many part of the relationship, in the rental class you did `the 'belongs-to' association` part of the many to many relationships. On getters and setters, I agree the reviewer might have used the terms wrongly, my apologies on behalf of the reviewer.

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @danielamoreno699 , 

Great work on making the changes requested by a previous reviewer 👏  👏 👏
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
✔️  Good work adding the association,` belongs_to` and `has_many` asssociations

 # Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by 1GORDON — APPROVED** [review]
# Status: APPROVED 🟢 

Hi @danielamoreno699 ✋

Your project is complete!  There is nothing else to say other than... it's time to merge it: ship it:
Congratulations! :tada: :clap: :clap: :100:

## Highlights!
- Associations well setup✔️

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend considering them as they can improve your code._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread and tag me `@1GORDON` if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**